Client doesn't start

```console
$ lein run client
Exception in thread "main" java.lang.NullPointerException, compiling:(T:\Temp\form-init7564624643793564877.clj:1:78)
        at clojure.lang.Compiler.load(Compiler.java:7391)
        at clojure.lang.Compiler.loadFile(Compiler.java:7317)
        at clojure.main$load_script.invokeStatic(main.clj:275)
        at clojure.main$init_opt.invokeStatic(main.clj:277)
        at clojure.main$init_opt.invoke(main.clj:277)
        at clojure.main$initialize.invokeStatic(main.clj:308)
        at clojure.main$null_opt.invokeStatic(main.clj:342)
        at clojure.main$null_opt.invoke(main.clj:339)
        at clojure.main$main.invokeStatic(main.clj:421)
        at clojure.main$main.doInvoke(main.clj:384)
        at clojure.lang.RestFn.invoke(RestFn.java:421)
        at clojure.lang.Var.invoke(Var.java:383)
        at clojure.lang.AFn.applyToHelper(AFn.java:156)
        at clojure.lang.Var.applyTo(Var.java:700)
        at clojure.main.main(main.java:37)
Caused by: java.lang.NullPointerException
        at clojure.lang.Numbers.ops(Numbers.java:1013)
        at clojure.lang.Numbers.multiply(Numbers.java:148)
        at clojure.lang.Numbers.multiply(Numbers.java:3740)
        at hyperspace.client.ui$render_player.invokeStatic(ui.clj:98)
        at hyperspace.client.ui$render_player.invoke(ui.clj:88)
        at hyperspace.client.ui$render_world.invokeStatic(ui.clj:117)
        at hyperspace.client.ui$render_world.invoke(ui.clj:105)
        at hyperspace.client.ui$ui_loop.invokeStatic(ui.clj:148)
        at hyperspace.client.ui$ui_loop.invoke(ui.clj:136)
        at hyperspace.client.ui$start.invokeStatic(ui.clj:157)
        at hyperspace.client.ui$start.invoke(ui.clj:154)
        at hyperspace.client.main$run.invokeStatic(main.clj:6)
        at hyperspace.client.main$run.doInvoke(main.clj:5)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.core$apply.invokeStatic(core.clj:646)
        at clojure.core$apply.invoke(core.clj:641)
        at hyperspace.main$_main.invokeStatic(main.clj:7)
        at hyperspace.main$_main.doInvoke(main.clj:5)
        at clojure.lang.RestFn.invoke(RestFn.java:410)
        at clojure.lang.Var.invoke(Var.java:379)
        at user$eval5.invokeStatic(form-init7564624643793564877.clj:1)
        at user$eval5.invoke(form-init7564624643793564877.clj:1)
        at clojure.lang.Compiler.eval(Compiler.java:6927)
        at clojure.lang.Compiler.eval(Compiler.java:6917)
        at clojure.lang.Compiler.load(Compiler.java:7379)
        ... 14 more
```

--- a/LICENSE.md
+++ b/LICENSE.md
@@ -1,7 +1,7 @@
 The MIT License
 ===============
 
-Copyright (C) 2012-2016 by hyperspace developers
+Copyright (C) 2012-2017 by hyperspace developers
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal

--- a/README.md
+++ b/README.md
@@ -7,7 +7,13 @@ Hyperspace is a [Slingshot](http://slingshot.wikispot.org/) clone written in Clo
 The game is under heavy development. Nothing works here. Move away!
 
 ## Game client
-Use `lein run client` to run the standalone game client.
+
+To run the standalone game client use one of the following commands:
+
+```console
+$ lein run client
+$ lein run
+```
 
 ## Game server
 

--- a/src/hyperspace/client/richworld.clj
+++ b/src/hyperspace/client/richworld.clj
@@ -18,6 +18,18 @@
   (assoc (world/make-missile position velocity)
     :trace-index trace-index))
 
+(defn add-missile
+  [{missiles :missiles
+    traces   :traces
+    :as world}
+   position velocity]
+  (let [new-missile (make-missile position
+                                  velocity
+                                  (count traces))]
+    (assoc world
+      :missiles (conj missiles new-missile)
+      :traces (conj traces []))))
+
 ;;; World-related stuff:
 
 (defn enrich-world
@@ -56,3 +68,20 @@
           :fragments new-fragments
           :traces new-traces)
         (- delta-time simulation/simulation-step)))))
+
+(defn fire
+  [{[player & _] :players
+    :as world}]
+  (let [{position               :position
+         [_, power :as heading] :heading} player
+        missile-position (-> (polar->cartesian heading)
+                             normalize-vector
+                             (multiply-by-scalar (+ world/player-radius
+                                                    world/missile-radius))
+                             (vector-sum position))
+        missile-velocity (-> (polar->cartesian heading)
+                             normalize-vector
+                             (multiply-by-scalar (* power 100)))]
+    (add-missile world
+                 missile-position
+                 missile-velocity)))

--- a/src/hyperspace/client/richworld.clj
+++ b/src/hyperspace/client/richworld.clj
@@ -23,7 +23,9 @@
 (defn enrich-world
   "Update player definitions in world (enrich with UI-required properties)."
   [world]
-  (update-in world [:players] (fn [players] (map enrich-player players))))
+  (update-in world [:players] (fn [players] (->> players
+                                                 (map enrich-player)
+                                                 (into [])))))
 
 (defn update-world
   [{missiles :missiles

--- a/src/hyperspace/client/richworld.clj
+++ b/src/hyperspace/client/richworld.clj
@@ -4,26 +4,26 @@
   (:use [hyperspace.client particles]
         [hyperspace.library geometry]))
 
-;;; Missles related stuff
+;;; Player-related stuff:
+
+(defn enrich-player
+  "Enrich player with UI-required properties."
+  [player]
+  (assoc player :heading [0 1]))
+
+;;; Missiles-related stuff:
 
 (defn make-missile
   [position velocity trace-index]
   (assoc (world/make-missile position velocity)
     :trace-index trace-index))
 
-(defn add-missile
-  [{missiles :missiles
-    traces   :traces
-    :as world}
-   position velocity]
-  (let [new-missile (make-missile position
-    velocity
-    (count traces))]
-    (assoc world
-      :missiles (conj missiles new-missile)
-      :traces (conj traces []))))
+;;; World-related stuff:
 
-;;; World updating stuff:
+(defn enrich-world
+  "Update player definitions in world (enrich with UI-required properties)."
+  [world]
+  (update-in world [:players] (fn [players] (map enrich-player players))))
 
 (defn update-world
   [{missiles :missiles
@@ -35,17 +35,17 @@
   (if (<= delta-time simulation/simulation-step)
     [world delta-time]
     (let [broken-particles (mapcat #(break-particle % planets)
-      (concat missiles fragments))
+                            (concat missiles fragments))
           ;; FIXME: Duplicate code
           new-missiles (->> missiles
-        (filter #(circle-X-rectangle? % world))
-        (remove #(circle-X-any-circle % planets))
-        (map #(simulation/update-particle % planets simulation/simulation-step)))
+                        (filter #(circle-X-rectangle? % world))
+                        (remove #(circle-X-any-circle % planets))
+                        (map #(simulation/update-particle % planets simulation/simulation-step)))
           new-fragments (->> fragments
-        (concat broken-particles)
-        (filter #(circle-X-rectangle? % world))
-        (remove #(circle-X-any-circle % planets))
-        (map #(simulation/update-particle % planets simulation/simulation-step)))
+                         (concat broken-particles)
+                         (filter #(circle-X-rectangle? % world))
+                         (remove #(circle-X-any-circle % planets))
+                         (map #(simulation/update-particle % planets simulation/simulation-step)))
 
           new-traces  (reduce update-traces traces new-missiles)]
       (recur

--- a/src/hyperspace/client/ui.clj
+++ b/src/hyperspace/client/ui.clj
@@ -153,5 +153,6 @@
 
 (defn start
   [world]
-  (setup-ui world)
-  (ui-loop world))
+  (let [world (enrich-world world)]
+    (setup-ui world)
+    (ui-loop world)))

--- a/src/hyperspace/client/ui.clj
+++ b/src/hyperspace/client/ui.clj
@@ -2,9 +2,10 @@
   (:import [org.lwjgl Sys]
            [org.lwjgl.opengl Display DisplayMode GL11]
            [org.lwjgl.input Mouse Keyboard])
-  (:use [hyperspace.library geometry simulation]
-        [hyperspace.client controls richworld])
-  (:require [hyperspace.library.world :as world]))
+  (:use [hyperspace.library geometry]
+        [hyperspace.client controls])
+  (:require [hyperspace.client.richworld :as richworld]
+            [hyperspace.library.world :as world]))
 
 (def window-width 800)
 (def window-height 600)
@@ -47,7 +48,7 @@
 
         (and (= key Keyboard/KEY_SPACE)
              (not (Keyboard/isRepeatEvent)))
-        (fire world)
+        (richworld/fire world)
 
         :else world))
     world))
@@ -144,7 +145,7 @@
 
       (let [new-timestamp (get-time)
             delta-time    (+ (- new-timestamp initial-timestamp) accumulated-time)
-            [new-world remaining-time] (update-world world delta-time)
+            [new-world remaining-time] (richworld/update-world world delta-time)
             final-world (-> new-world process-input
                                       render-world)]
         (Display/update)
@@ -153,6 +154,6 @@
 
 (defn start
   [world]
-  (let [world (enrich-world world)]
+  (let [world (richworld/enrich-world world)]
     (setup-ui world)
     (ui-loop world)))

--- a/src/hyperspace/main.clj
+++ b/src/hyperspace/main.clj
@@ -2,7 +2,9 @@
   (:require [hyperspace.client.main :as client]
             [hyperspace.server.server :as server]))
 
-(defn -main [application & args]
-  (cond
-    (= application "client") (apply client/run args)
-    (= application "server") (apply server/start args)))
+(defn -main [& args]
+  (let [[application & app-args] args]
+    (cond
+      (= application "client") (apply client/run app-args)
+      (= application "server") (apply server/start app-args)
+      :else (apply client/run args))))

--- a/test/hyperspace/test/client/controls.clj
+++ b/test/hyperspace/test/client/controls.clj
@@ -1,0 +1,11 @@
+(ns hyperspace.test.client.controls
+  (:use [hyperspace.client.controls]
+        [hyperspace.client.richworld]
+        [hyperspace.test.checkers]
+        [midje.sweet]))
+
+(let [world (enrich-world {:players [{}]})]
+  (facts "about turn-left function"
+         (turn-left world 5) => {:players [{:heading [5 1]}]})
+  (facts "about turn-right function"
+         (turn-right world 5) => {:players [{:heading [-5 1]}]}))

--- a/test/hyperspace/test/client/richworld.clj
+++ b/test/hyperspace/test/client/richworld.clj
@@ -5,3 +5,9 @@
 
 (facts "about make-missile function"
   (:trace-index (make-missile [10 20] [30 40] 888)) => 888)
+
+(facts "about enrich-player function"
+       (enrich-player {}) => {:heading [0 1]})
+
+(facts "about enrich-world function"
+       (enrich-world {:players [{}]}) => {:players [{:heading [0 1]}]})

